src/lib/vrf_evaluator/vrf_evaluator.ml

```diff
@@ -0,0 +1,408 @@
+            Queue.enqueue slots_won slot_won ;
+            find_winning_slot (Consensus_time.succ consensus_time)
+    in
+    find_winning_slot start_consensus_time
+
+  let create config =
+    let reset_reader, reset_writer =
+      Pipe_lib.Strict_pipe.create ~name:"Vrf-evaluator reset"
+        Pipe_lib.Strict_pipe.Synchronous
+    in
+    let t =
+      { config
+      ; last_checked_slot_and_epoch= Public_key.Compressed.Table.create ()
+      ; slots_won=
+          Queue.create
+            ~capacity:
+              (Global_slot.to_int config.consensus_constants.slots_per_epoch)
+            ()
+      ; current_slot= None
+      ; current_epoch= Epoch.zero
+      ; epoch_state= {block_producer_keys= []}
+      ; reset_writer }
+    in
+    don't_wait_for
+      (Strict_pipe.Reader.iter_without_pushback reset_reader
+         ~f:(fun epoch_data ->
+           [%log' info config.logger] "Resetting epoch data for epoch %d"
+             (Epoch.to_int epoch_data.epoch) ;
+           Queue.clear t.slots_won ;
+           t.current_slot <- None ;
+           t.current_epoch <- epoch_data.epoch ;
+           don't_wait_for (evaluate epoch_data t) )) ;
+    return t
+end
+
+module Functions = struct
+  type ('i, 'o) t =
+    'i Bin_prot.Type_class.t
+    * 'o Bin_prot.Type_class.t
+    * (Worker_state.t -> 'i -> 'o Deferred.t)
+
+  let create input output f : ('i, 'o) t = (input, output, f)
+
+  let set_new_epoch_state =
+    create Consensus.Data.Epoch_data_for_vrf.Stable.Latest.bin_t Unit.bin_t
+      (fun w e ->
+        let logger = w.config.logger in
+        let update () =
+          [%log info]
+            "Updating epoch data for the VRF evaluation for epoch $epoch"
+            ~metadata:[("epoch", Epoch.to_yojson e.epoch)] ;
+          Strict_pipe.Writer.write w.reset_writer e
+        in
+        if Epoch.(succ e.epoch > w.current_epoch) then (
+          [%log info] "Updating epoch data" ;
+          update () )
+        else (
+          [%log info] "Received epoch data for current epoch $epoch. Skipping "
+            ~metadata:[("epoch", Epoch.to_yojson e.epoch)] ;
+          Deferred.unit ) )
+
+  let slots_won_so_far =
+    create Unit.bin_t Vrf_evaluation_result.Stable.Latest.bin_t (fun w () ->
+        let slots_won = Queue.to_list w.slots_won in
+        [%log' info w.config.logger]
+          !"Slots won evaluator: %{sexp: Consensus.Data.Slot_won.t list}"
+          slots_won ;
+        Queue.clear w.slots_won ;
+        let evaluator_status =
+          match w.current_slot with
+          | Some slot ->
+              Evaluator_status.At slot
+          | None ->
+              Completed
+        in
+        return Vrf_evaluation_result.{slots_won; evaluator_status} )
+
+  let update_block_producer_keys =
+    create State.Stable.Latest.bin_t Unit.bin_t (fun w e ->
+        let logger = w.config.logger in
+        [%log info] "Updating block producer keys" ;
+        w.epoch_state <- e ;
+        Deferred.unit )
```

Comment:
Call `don't_wait_for (evaluate epoch_data t)` here and interrupt the currently running one.